Spike: Add ThrottledCall utility class

citizen428 opened the pull request:
## What type of PR is this? (check all applicable)

- [X] Feature

## Description

Based on our discussions in #3649, this PR implements an MVP version of a `ThrottledCall` utility class, which will execute a block and then prevent its repeated execution for a specified time interval:

```ruby
# This will execute the block and sleep for 3 seconds
ThrottledCall.call(:example, throttle_for: 5.minutes) { sleep 3 }

# This immediately returns true and won't execute the block
ThrottledCall.call(:example, throttle_for: 5.minutes) { sleep 3 }
```

Given that this is a spike the utility class is basically just a very thin wrapper around `Rails.cache.fetch` which already handles all the important logic, TTL, etc. The wrapper just adds key namespacing (all keys are prefixed with `throttled_call.`) and normalizes the stored value to `true` instead of the block result. I also experimented with returning the TTL (`Rails.cache.redis.ttl(namespaced_key)`) but I didn't think this provided enough benefit to be worth the extra roundtrip.

## Related Tickets & Documents

#3649

## QA Instructions, Screenshots, Recordings

If you want to try this out in your local environment, you need to activate the Redis cache store:

```sh
rails dev:cache
```

### UI accessibility concerns?

n/a

## Added tests?

- [X] Yes

rhymes commented:
Cool example! I'm going to play a bit with it and report back 

> I also experimented with returning the TTL (Rails.cache.redis.ttl(namespaced_key)) but I didn't think this provided enough benefit to be worth the extra roundtrip.

These should addressable using `MULTI`: https://github.com/redis/redis-rb#executing-commands-atomically

rhymes reviewed (commented):
I played a bit with it and the only thing I find odd is that the key name is decoupled from the action itself but that's probably inevitable. The only "risk" is that we use the same key in different parts of the code resulting in the cache to be overwritten with a different result.

For a second I entertained an idea like https://github.com/Shopify/limiter/blob/3f30f40e8b5fc6fc98e3a82ef13f01a9800f30ce/lib/limiter/mixin.rb but that hides the rate limiting moving it down where the method is defined and complicates things a lot (also because we might want to rate limit the same method call differently in different parts of the code)

I'd love to see it in action in the spike replacing one of the `rand()` calls.

mstruve reviewed (commented):
I like it! Let's see it in action :)

citizen428 commented:
> These should addressable using 

Apart from not being accessible from `Rails.cache` I don't think who this would work with Ruby code in between the two Redis commands.

> the only thing I find odd is that the key name is decoupled from the action itself 

Agreed, but it's simple and I guess we would generally use better key names than 'test' or 'example'.

> For a second I entertained an idea like

The first version I started prototyping in my head was a mix-in. I decided against it because I like the explicit nature of this utility class, but it did remove the option of using e.g. `__method__` to derive a key name. There are still ways we could make keys more unique (e.g. by hashing the block object), but I think we can do that later if we need it.

citizen428 commented on app/workers/reactions/update_relevant_scores_worker.rb:
```diff
@@ -9,7 +9,11 @@ def perform(reaction_id)
       return unless reaction&.reactable
 
       reaction.reactable.touch_by_reaction if reaction.reactable.respond_to?(:touch_by_reaction)
-      reaction.reactable.sync_reactions_count if rand(6) == 1 && reaction.reactable.respond_to?(:sync_reactions_count)
```
There actually don't seem to be too many places left where we use these randomized calls, but here is one of them. The main problem is that I completely guessed how long we throttle for (please let me know if you'd prefer a value other than 15 minutes), whereas the `rand` based approach just says "do this roughly one out of 6 times".

rhymes replied:
I'm quite sure these numbers are based on observation of past performance woes in production so unless we literally benchmark this in prod and compare the before and after, we can rely on our collective judgments for the new intervals

mstruve replied:
Knowing when to bust a cache is always hard and I have found that usually, you have to make your best guess at what will work and then be willing to tweak as necessary. 

That aside, at least we have something concrete to tweak and not a `rand` 🙈

citizen428 commented:
@rhymes and @mstruve I now changed the class method from `call` to `perform`, since `ThrotteldCall.call` sounded odd. I also added an example of how using this would look.

rhymes reviewed (commented):
Thanks for the update!

mstruve commented on app/workers/reactions/update_relevant_scores_worker.rb:
```diff
@@ -9,7 +9,11 @@ def perform(reaction_id)
       reaction.reactable.touch_by_reaction if reaction.reactable.respond_to?(:touch_by_reaction)
-      reaction.reactable.sync_reactions_count if rand(6) == 1 && reaction.reactable.respond_to?(:sync_reactions_count)
+      if reaction.reactable.respond_to?(:sync_reactions_count)
+        ThrottledCall.perform(:sync_reactions_count, throttle_for: 15.minutes) do
```
Should we throw a test in for this?

citizen428 replied:
My plan has been foiled... 😉 

![image](https://user-images.githubusercontent.com/47985/109120355-6d644200-7778-11eb-9269-72b3243d3cc9.png)

mstruve reviewed (approved):
Might want to add a spec but code looks solid and feels SO MUCH better. 🚀

citizen428 commented on spec/workers/reactions/update_relevant_scores_worker_spec.rb:
```diff
@@ -45,5 +45,10 @@
       end.not_to raise_error
     end
+
+    it "uses a throttled call for syncing the reactions count" do
```
@mstruve Added this spec now. I think since `ThrottledCall` has unit tests, it's enough to check that it receives the right arguments. I didn't want the spec to fail every time we update the interval, so I only verify the class of the second argument.

mstruve replied:
Perfect! Thanks!

rhymes reviewed (approved):
I think we could go forward with this, one final thing: how do we know this is actually working?

I'm not saying the `ThrottledCall` object which obviously does what it's meant to do, but how do we know we've successfully replaced a `rand()` call with `ThrottledCall` as we have no visibility in it?

Should we collect some low key stats about `HIT/MISS` rates?

mstruve commented:
> how do we know this is actually working?

Once it's deployed I can go in a console and check for the Redis key. Anyone can do with on the Ben halpern community. Might have to go around doing things to update the scores first ;) We can also go into Honeycomb, find the update relevant scores worker and look at the traces to ensure that its not executing the update on every single run.

citizen428 commented:
@rhymes I'm going to merge and deploy this now. I think what @mstruve says makes sense for now, if we need more instrumentation or something we can make a 15-minute fix PR for it.
